### ISSUE TITLE
Define a better load balancing strategy

### ISSUE THREAD

**Issue by heurtematte** [eclipse-cbi/buildkitd-okd#3]
Building docker images at the same time using the current buildkit infra using a kubernetes service and ClusterIP often finishes with an OOM, a
error: failed to solve: Canceled: context canceled, or another issue due to lack of memory. 

For instance, after multiple executions, kubernetes only once decided to use a different pod because I manually delayed the job execution. 
Otherwise kubernetes always choose the same pod. which seems normal, because when jobs start they don't use many resources.

![image](https://github.com/eclipse-cbi/buildkitd-okd/assets/8045662/f9a34850-b34f-4833-978a-531d08488c8e)

Is there a way to change the service load balancing strategy in OKD? I haven't found anything.
Or maybe create an OKD route by defining a `leastconn` strategy in `haproxy.router.openshift.io/balance` annotation, see: https://docs.okd.io/latest/networking/routes/route-configuration.html, but buildkit is TCP, and there is an open issue about exposing buildkit behind HTTP: https://docs.okd.io/latest/networking/routes/route-configuration.html

Any idea to better distribute the load?

**Comment by heurtematte** [comment]
@fredg02 @mbarbero

**Comment by mbarbero** [comment]
Routes (or nodeports) are ways to expose a service externally to the cluster, so going this way would not change anything about service load balancing.

ClusterIP services load balancing is random:

```
❯ kubectl get service buildkitd -n foundation-internal-infra-buildkitd
NAME        TYPE        CLUSTER-IP      EXTERNAL-IP   PORT(S)    AGE
buildkitd   ClusterIP   A.B.C.D   <none>        1234/TCP   2y109d
```

Note the cluster IP `A.B.C.D`. Now, let's log into a node 

```
❯ oc debug node/okdnode-XXX
sh-4.4# chroot /host
sh-5.2# iptables -n -t nat -L | grep A.B.C.D
KUBE-SVC-QWERTYUIOP  6    --  0.0.0.0/0            A.B.C.D        /* foundation-internal-infra-buildkitd/buildkitd cluster IP */ tcp dpt:1234
KUBE-MARK-MASQ  6    --  0.0.0.0/0            A.B.C.D        /* foundation-internal-infra-buildkitd/buildkitd cluster IP */ tcp dpt:1234
sh-5.2# iptables -n -t nat -L KUBE-SVC-QWERTYUIOP
Chain KUBE-SVC-QWERTYUIOP (1 references)
target     prot opt source               destination
KUBE-MARK-MASQ  6    --  0.0.0.0/0            A.B.C.D        /* foundation-internal-infra-buildkitd/buildkitd cluster IP */ tcp dpt:1234
KUBE-SEP-XXXXXXXXXXXXXXXX  0    --  0.0.0.0/0            0.0.0.0/0            statistic mode random probability 0.20000000019
KUBE-SEP-YYYYYYYYYYYYYYYY  0    --  0.0.0.0/0            0.0.0.0/0            statistic mode random probability 0.25000000000
KUBE-SEP-ZZZZZZZZZZZZZZZZ  0    --  0.0.0.0/0            0.0.0.0/0            statistic mode random probability 0.33333333349
KUBE-SEP-WWWWWWWWWWWWWWWW  0    --  0.0.0.0/0            0.0.0.0/0            statistic mode random probability 0.50000000000
KUBE-SEP-UUUUUUUUUUUUUUUU  0    --  0.0.0.0/0            0.0.0.0/0
sh-5.2# iptables -n -t nat -L KUBE-SEP-ZZZZZZZZZZZZZZZZ
Chain KUBE-SEP-ZZZZZZZZZZZZZZZZ (1 references)
target     prot opt source               destination
KUBE-MARK-MASQ  0    --  J.K.L.M          0.0.0.0/0
DNAT       6    --  0.0.0.0/0            0.0.0.0/0            tcp to:J.K.L.M:1234
```

You see that the iptable load balancing is equally distributed. See https://github.com/kubernetes/kubernetes/issues/37932#issuecomment-264950788 if the increasing probability is challenging this thought. Each `KUBE-SEP` rule points to a pod's IP

There is probably connection tracking happening somewhere. Could you please try to run a new build, and once you know on which node the build is scheduled (and once a first container is being build), run:

```
oc debug "node/okdnode-XXX" -- sh -c 'cat /proc/net/nf_conntrack | grep A.B.C.D'
```

This will help us identifying if conntrack is at play here (which is my best guess)

**Comment by heurtematte** [comment]
So as discussed there is no easy way to solve this problem, except by adding more resources.

For now, I prefer to avoid this last option, I will document in the CBI wiki a recommendation like building sequentially or delaying by a few-second container build when they are resource-consuming and running in parallel.

**Comment by mbarbero** [comment]
Ensuring that `memory.limits` matches `memory.requests` can significantly enhance the stability of the workload. This prevents the pod from being evicted by the scheduler invoking the OOMKiller when there's increased memory pressure on the node it runs on. I'd suggest trying that first.
